sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -495,6 +495,23 @@ object ColumnPruning extends Rule[LogicalPlan] {
     // Eliminate no-op Projects
     case p @ Project(_, child) if sameOutput(child.output, p.output) => child
 
+    // The column of father project contains not deterministic function
+    // e.g Rand function. father project will be split to two project.
+    case h @ Project(fields, _: LeafNode) if !fields.forall(_.deterministic) =>
```

Comment:
It was the way I tested it. otherwise PhysicalOperation  fails for non-deterministic， hive read all columns from the table.
Of course， for the case like the following:
```
Project [rand()]
   TableScan XXX_table
```
PhysicalOperation still fails for non-deterministic.  So we still read all columns from the table.
Because I don't know how to read that column.
Or we need to read all the columns in advance to select one column, but I'm not sure if this will work.